docs/design/partitioning/07-parallel-migrations.md

```diff
@@ -0,0 +1,536 @@
+| Partition Table Length      |
++-----------------------------+
+| Replicas (Address + UUID)   |
++-----------------------------+
+| Partition-0 Version         |
++-----------------------------+
+| Partition-0 Replica Indexes |
++-----------------------------+
+| Partition-1 Version         |
++-----------------------------+
+| Partition-1 Replica Indexes |
++-----------------------------+
+| ........................... |
++-----------------------------+
+| Partition-N Version         |
++-----------------------------+
+| Partition-N Replica Indexes |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+     
+```          
+
+Second change is in the partition table restore process. While restoring partition table, if a replica address change is
+detected, we replace the old address of the replica with the new one in the partition table. In current mechanism, 
+global partition version is incremented by the number of updates, as if partition table is updated in the runtime.
+With the new versioning, only updated partitions' versions will be incremented.
+
+Third change is in the partial restart decision process. During partial restart, we decide which members to perform the 
+restart, which members to be excluded and delete their data. To select member groups having the same partition table,
+we use the partition table version. Members with the same partition table are grouped together. Also to find the most 
+recent partition table, we pick the partition table with the greatest version. If the policy is `PARTIAL_RECOVERY_MOST_RECENT`,
+we pick the group the greatest partition table version. If it's `PARTIAL_RECOVERY_MOST_COMPLETE`, then we pick 
+the largest group. In other places we will use _partition table stamp_ instead of removed partition table version. But in
+this case, we cannot use _partition table stamp_. Because stamp is only valid for equality comparison, it cannot be used
+to compare order of the partition tables. Instead we will use the sum of per-partition versions to compare the order. 
+_This works because Hot Restart does not support arbitrary server crashes. Only proper server shutdowns are supported._
+Otherwise, resolving conflicts on partition level, determining the most recent versions of each partition, composing
+a new partition table with these partitions and adapting Hot Restart to restore data based on this new partition table
+requires a large rework and time. And supporting rolling upgrade scenarios makes this more complicated.            
+
+
+#### 3.3 - Rolling Upgrade
+
+- To support rolling upgrade, we will have two distinct paths for most the migration scheduling and execution logic. 
+
+- While updating partition table and applying completed migrations, there'll be separate paths too.
+
+- Once the cluster is upgraded to `4.1`, all partition versions will be set to the latest partition table version.
+
+- Cluster state change transaction mechanism will use either partition table version or stamp according to the version.
```

Comment:
instead of `according to the version`, rewording it as `according to the cluster version` seemed clearer to understand.